apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/MetricRegistry.java

```diff
@@ -43,6 +45,8 @@ public class MetricRegistry {
     private static final Logger logger = LoggerFactory.getLogger(MetricRegistry.class);
     private final WriterReaderPhaser phaser = new WriterReaderPhaser();
     private final ReporterConfiguration reporterConfiguration;
+
+    private final Set<MetricsProvider> metricsProviders = Collections.newSetFromMap(new ConcurrentHashMap<MetricsProvider, Boolean>());
```

Comment:
what limits the size of this set?

Reply:
`MetricsProvider` are very much like our gauges: There's a finite set defined by our agent implementation (=all calls to `MetricRegistry.addMetricsProvider`). Therefore there is no risk of this set growing out of control.